runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -1873,14 +1915,20 @@ private void sendWorkerUpdatesToDataflowService(
    */
   private void refreshActiveWork() {
     Map<String, List<Windmill.KeyedGetDataRequest>> active = new HashMap<>();
+    Map<String, List<Windmill.HeartbeatRequest>> heartbeats = new HashMap<>();
     Instant refreshDeadline =
         clock.get().minus(Duration.millis(options.getActiveWorkRefreshPeriodMillis()));
 
     for (Map.Entry<String, ComputationState> entry : computationMap.entrySet()) {
-      active.put(entry.getKey(), entry.getValue().getKeysToRefresh(refreshDeadline, sampler));
+      if (windmillServiceEnabled
+          && DataflowRunner.hasExperiment(options, "send_new_heartbeat_requests")) {
+        heartbeats.put(entry.getKey(), entry.getValue().getKeyHeartbeats(refreshDeadline, sampler));
+      } else {
+        active.put(entry.getKey(), entry.getValue().getKeysToRefresh(refreshDeadline, sampler));
```

Comment:
Yeah. My understanding is that Windmill can't handle requests that contain both the old and new style at the same time.

Reply:
since it is the same information, can the switch on which type of request is sent to windmill be moved into the actual rpc layer?  Then you could just use the heartbeat objects internally and covert there if neeeded.

I think that woudl be simpler than the somewhat duplicated methods and params and keeps it more self-contained.

Reply:
Are you suggesting we plumb the `computationMap` down into the RPC layer instead? Otherwise, I could build a map of consisting of a new type of object (maybe ComputationGetDataRequest since that contains both types of heartbeats or maybe a new object entirely) and pass that down into the RPC layer. In one case we need the user key and in the other we need the cache token, but the rest of the data is the same between the two.

Reply:
I don't want to plumb computationMap as it is rather complex. I forgot the that KeyedGetDataRequest doesn't have the cache_token.

So I think you could switch to new heartbeat format at this level and pass that down to rpc layer. And it can convert back to legacy heartbeats format if it is configured to.  I think that in both previous and new heartbeat formats we don't want to send the user `key` as it can be large and cause us to exceed the request size limit and is unnecessary in windmill.  So the new proto should have sufficient information for both cases.

Reply:
It's adding a bit more to the implementation, but I've changed the worker to not add the user key to the heartbeats. Note I needed to change windmill.proto to have the user key be optional instead of required, but I think this should be compatible since Windmill is ignoring the field anyway (and the internal version of the proto already seems to have it as optional as well).

Reply:
Sounds good, v2 runner also doesn't set the key in this path as another confirmation this works.